### PR TITLE
refactor(webserver): extract RegisterError and TokenAuthError and add unit test

### PR DESCRIPTION
**Pull request by wsxiaoys**

### CODE REVIEW

File: ee/tabby-webserver/src/service/db.rs [hunk 1]
```diff
@@ -444,14 +457,44 @@ mod tests {
...
+        create_admin_user(&conn).await;
+
+        assert!(is_enum_variant!(
+            conn.token_auth(ADMIN_EMAIL.to_owned(), "12345678".to_owned())
+                .await
+                .unwrap_err(),
+            TokenAuthError::InvalidPassword
+        ));
+
+        // This won't work, due to password hash is not right.
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/db.rs:492]: @darknight

File: ee/tabby-webserver/src/service/auth.rs [hunk 2]
```diff
@@ -263,4 +274,105 @@ mod tests {
...
+        create_admin_user(&conn).await;
+
+        assert_matches!(
+            conn.token_auth(ADMIN_EMAIL.to_owned(), "12345678".to_owned())
+                .await
+                .unwrap_err(),
+            TokenAuthError::InvalidPassword
+        );
+
+        // This won't work, due to password hash is not right.
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/auth.rs:306]: @darknight

File: ee/tabby-webserver/src/service/auth.rs [hunk 3]
```diff
@@ -263,4 +270,105 @@ mod tests {
         assert!(password_verify(raw, &hash));
         assert!(!password_verify(raw, "invalid hash"));
     }
+
+    static ADMIN_EMAIL: &str = "test@example.com";
+    static ADMIN_PASSWORD: &str = "123456789";
+
+    async fn create_admin_user(conn: &DbConn) -> i32 {
+        conn.create_user(ADMIN_EMAIL.to_string(), ADMIN_PASSWORD.to_string(), true)
```
Comment (darknight) [inline ee/tabby-webserver/src/service/auth.rs:278]: password is saved as plain text, not hash.

```
conn.create_user(ADMIN_EMAIL.to_string(), password_hash(ADMIN_PASSWORD).unwrap(), true)
```
fixes the broken test.

File: ee/tabby-webserver/src/schema/auth.rs [hunk 4]
```diff
@@ -143,11 +201,17 @@ pub trait AuthenticationService: Send + Sync {
...
+
+    async fn token_auth(
+        &self,
+        email: String,
+        password: String,
+    ) -> std::result::Result<TokenAuthResponse, TokenAuthError>;
+
+    async fn refresh_token(&self, refresh_token: String) -> Result<RefreshTokenResponse>;
+    async fn verify_token(&self, access_token: String) -> Result<VerifyTokenResponse>;
+    async fn is_admin_initialized(&self) -> Result<bool>;
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/auth.rs:214]: rule of thumb:
1. By default, starts with anyhow::Result.
2. If there's a need for Validation / Other error type, define a new error and iterate.

Sounds like a good iterative approach?

cc @darknight

**Pull request merged**